github/commitstatus.go

```diff
@@ -225,17 +227,19 @@ func min(a, b int) int {
 }
 
 // checkForRetry determines if we should retry the http request and calculates wait time between retries
-func checkForRetry(res httpResponse) time.Duration {
+func checkForRetry(res httpResponse, waitTime, maxSleepTime time.Duration) time.Duration {
 	switch {
 	// If you exceed the rate limit, the response will have a 403 status and the x-ratelimit-remaining header will be 0
 	// https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit
 	case res.statusCode == http.StatusForbidden && res.rateLimitRemaining == 0:
 		sleepTime := time.Until(res.rateLimitReset)
-		if sleepTime > MaxSleepTime {
+		if sleepTime > maxSleepTime {
 			return 0
 		} else {
 			return sleepTime
 		}
+	case res.statusCode >= 500:
```

Comment:
I just want to make it loud and clear, the basic rule is: do not retry on 5xx, unless you understand what you are doing.

The 5xx category of error is meant for cases when you, as a client, cannot know the server's state. So as a general rule one must assume that it is not safe to retry a request that failed with an error code in that category.

For example, take the typical 504 Gateway Timeout, it means the server took so long to process the request that the gateway gave up before seeing the answer. As a client you don't know if it processed it successfully or not. If you blindly retry it might happen that the request is successfully executed each time and if that request was to pay for something, you might end up paying N times instead of 1. But it might also fail on the second call, for instance because the transaction id was consumed in the previous (successful) call, and seeing the request failed the user might want to retry from the beginning thus purchasing something twice. This should show that it is always tricky to interpret 5xx and decide whether to retry or not unless you get a good understanding of what's going on.

Really keep in mind that whenever you want to retry an API call. Always ensure it is safe to do so for that particular API call.

In the case of cogito, it is safe to retry. At worst we'll be setting the commit status multiple times, which should be an idempotent operation anyway so that's safe.

Reply:
> This risks becoming overengineering in the wrong place (the resource) to solve a human problem (people putting 10 retries) ...

Regarding the `overengineering` part. The following lines are the only required change....
```
case res.statusCode >= 500:
   return waitTime
```

Reply:
In this case, the "overengineering" referred to the complexity in the behavior (this is a huge change), not to the complexity of the implementation.

Reply:
Regarding the retry on 5xx, I thought a bit more. I think it is dangerous to simply detect >= 500, because some of the 5xx errors are permanent (see for example the list in https://en.wikipedia.org/wiki/List_of_HTTP_status_codes). So I think that a conservative list is 500,502,503,504. For these, we retry, For the others, we fail.

Reply:
Done as suggested.